Sources/SwiftGodot/Core/GenericSignal.swift

```diff
@@ -0,0 +1,147 @@
+        if r != .ok { print("Warning, error connecting to signal, code: \(r)") }
+        return signalProxy
+    }
+
+    /// Disconnects a signal that was previously connected, the return value from calling
+    /// ``connect(flags:_:)``
+    public func disconnect(_ token: Object) {
+        target.disconnect(signal: signalName, callable: Callable(object: token, method: SignalProxy.proxyName))
+    }
+
+    /// You can await this property to wait for the signal to be emitted once.
+    public var emitted: Void {
+        get async {
+            await withCheckedContinuation { c in
+                let signalProxy = SignalProxy()
+                signalProxy.proxy = { _ in c.resume() }
+                let callable = Callable(object: signalProxy, method: SignalProxy.proxyName)
+                let r = target.connect(signal: signalName, callable: callable, flags: UInt32(Object.ConnectFlags.oneShot.rawValue))
+                if r != .ok { print("Warning, error connecting to signal, code: \(r)") }
+            }
+
+        }
+
+    }
+
+}
+
+/// This is a convenience object used as a helper for signals, all it does is
+/// register a method with godot called `proxy` that will invoke
+/// the callback defined in the public `proxy` variable here.
+///
+/// Usage:
+/// ```
+/// let demo = SignalProxy ()
+/// demo.proxy = { print ("called back") }
+///
+/// // Let godot call `demo` and its proxy method.
+/// invokeScript ("myDemo.proxy ()", params: ["myDemo", demo])
+/// ```
+public class SignalProxy: Object {
+    public static var proxyName = StringName("proxy")
+    static var initClass: Bool = {
+        register(type: SignalProxy.self)
+
+        let s = ClassInfo<SignalProxy>(name: "SignalProxy")
+
+        s.registerMethod(name: SignalProxy.proxyName, flags: .default, returnValue: nil, arguments: [], function: SignalProxy.proxyFunc)
+        return true
+    }()
+
+    /// The code invoked when Godot invokes the `proxy` method on this object.
+    public typealias Proxy = (borrowing Arguments) -> Void
+    public var proxy: Proxy?
+
+    public required init() {
+        let _ = SignalProxy.initClass
+        super.init()
+    }
+
+    public required init(nativeHandle: UnsafeRawPointer) {
+        super.init(nativeHandle: nativeHandle)
+    }
+
+    func proxyFunc(args: borrowing Arguments) -> Variant? {
+        proxy?(args)
+        return nil
+    }
+}
+
+/// The simple signal is used to raise signals that take no arguments and return no values.
+public typealias SimpleSignal = GenericSignal< /* no generic arguments */>
+
+extension Arguments {
+    enum UnpackError: Error {
+        case typeMismatch
+        case missingArgument
+    }
+
+    /// Unpack an argument as a specific type.
+    /// We throw a runtime error if the argument is not of the expected type,
+    /// or if there are not enough arguments to unpack.
+    func unpack<T: VariantStorable>(as type: T.Type, index: inout Int) throws -> T {
```

Comment:
I was hoping I could just call makeOrUnpack here, but the compiler can't decide between the two implementations and so says that they are ambiguous. 

I think that this is because T is an existential VariantStorable here, and so the compiler has lost any knowledge of whether it's also an Object.

Similarly I was hoping I could call `asObject(type)`, but the same problem exists - it complains that T needs to be an object.

I also tried specialising `func unpack<T: VariantStorable>(as type: T.Type, index: inout Int)` for T: Object, hoping that the pack expansion at the call site would pick the right version for each T -- but it doesn't seem to.

Someone smarter than me might be able to figure that stuff out, in which case this code could be a bit cleaner.

Reply:
Update: @migueldeicaza maybe @elijah-semyonov's variant changes will have made this easier; I haven't looked yet, but will try to.

Reply:
I've cleaned this up a bit, and added an explanatory comment about why we can't just call T.unpack.
I'm happy with this now.